File: examples/h2_stv.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "cint.h"

int cint1e_ovlp_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

int cint1e_kin_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

int cint1e_nuc_cart(double *buf, int *shls,
                    int *atm, int natm, int *bas, int nbas, double *env);

typedef int (* Int_cart)(double *buf, int *shls, 
					int *atm, int natm, int *bas, int nbas, double *env);

void read_arrays(FILE * file, int natm, int nbas, int ** atm, int ** bas, double ** env)
{
	double num;
	int number;
	int count = 0;

	while (count < natm * ATM_SLOTS) {
        fscanf(file, "%d", &number);
		(*atm)[count] = number;
		count++;
	}

	count = 0;
	while (count < nbas * BAS_SLOTS) {
        fscanf(file, "%d", &number);
		(*bas)[count] = number;
		count++;
	}

	count = 0;
	while (fscanf(file, "%lf", &num) != EOF) {
		(*env)[count] = num;
		count++;
	}
	fclose(file);
}

int main(int argc, char ** argv)
{	
	int natm = 2;
	int nbas = 2;

	int * atm = malloc(sizeof(int) * natm * ATM_SLOTS);
	int * bas = malloc(sizeof(int) * nbas * BAS_SLOTS);
	double * env = malloc(sizeof(double) * 10000);

	FILE * file = fopen("/u/jpmedina/libcint/molecules/h2/basis.txt", "r");
 	read_arrays(file, natm, nbas, &atm, &bas, &env);
	
	int di, dj;
	int shls[4];

	double * buf;
	
	printf("buf:\n");

	Int_cart func = cint1e_nuc_cart;

	for (int i = 0; i < nbas; i++) {
		for (int j = 0; j < nbas; j++) {
			shls[0] = i; di = CINTcgto_cart(i, bas);
			shls[1] = j; dj = CINTcgto_cart(j, bas);

			buf = malloc(sizeof(double) * di * dj);

			func(buf, shls, atm, natm, bas, nbas, env);

			for (int k = 0; k < di * dj; k++) {
				printf("%f ", buf[k]);
			}
			free(buf);
		}
		printf("\n");
	}
	printf("\n");

	free(atm);
	free(bas);
	free(env);
}

File: librint/grad/h2/h2_stv.rs
#![allow(non_snake_case, non_upper_case_globals)]

use std::io;

use librint::utils::read_basis;

use librint::cint_bas::CINTcgto_cart;
use librint::cint1e::cint1e_ovlp_cart;
use librint::cint1e::cint1e_nuc_cart;
use librint::intor1::cint1e_kin_cart;

pub const ATM_SLOTS: usize = 6;
pub const BAS_SLOTS: usize = 8;

fn main() -> io::Result<()>{
    let mut atm: Vec<i32> = Vec::new();
    let mut bas: Vec<i32> = Vec::new();
    let mut env: Vec<f64> = Vec::new();

    let path = "/u/jpmedina/libcint/molecules/h2o/sto3g.txt";
    read_basis(path, &mut atm, &mut bas, &mut env);

    const natm: usize = 2;
    const nbas: usize = 2;
    
    let mut shls: [i32; 4] = [0, 0, 0, 0];

    let mut buf;

	println!("ovlp");
    for i in 0..nbas {
        for j in 0..nbas {
            shls[0] = i as i32;
            shls[1] = j as i32;
            
            let di = CINTcgto_cart(i, &bas);
            let dj = CINTcgto_cart(j, &bas);

            buf = vec![0.0; (di * dj) as usize];
            cint1e_ovlp_cart(&mut buf, &mut shls, &mut atm, natm as i32, &mut bas, nbas as i32, &mut env);

            for i in 0..((di*dj) as usize) {
                print!("{} ", buf[i]);
            }
        }
        println!();
    }

    println!("kin");
    for i in 0..nbas {
        for j in 0..nbas {
            shls[0] = i as i32;
            shls[1] = j as i32;
            
            let di = CINTcgto_cart(i, &bas);
            let dj = CINTcgto_cart(j, &bas);

            buf = vec![0.0; (di * dj) as usize];
            cint1e_kin_cart(&mut buf, &mut shls, &mut atm, natm as i32, &mut bas, nbas as i32, &mut env);

            for i in 0..((di*dj) as usize) {
                print!("{} ", buf[i]);
            }
        }
        println!();
    }

    println!("nuc");
    for i in 0..nbas {
        for j in 0..nbas {
            shls[0] = i as i32;
            shls[1] = j as i32;
            
            let di = CINTcgto_cart(i, &bas);
            let dj = CINTcgto_cart(j, &bas);

            buf = vec![0.0; (di * dj) as usize];
            cint1e_nuc_cart(&mut buf, &mut shls, &mut atm, natm as i32, &mut bas, nbas as i32, &mut env);

            for i in 0..((di*dj) as usize) {
                print!("{} ", buf[i]);
            }
        }
        println!();
    }
    
    Ok(())
}